File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/infrastructure/ConversationContracts.java
package systems.zlink.samples.supportchat.server.support.infrastructure;

import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.server.support.domain.Conversation;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

public final class ConversationContracts {
    private ConversationContracts() {}

    public static Messages.ConversationState state(Conversation.Snapshot value) {
        return new Messages.ConversationState(
                value.conversationId(),
                value.subject(),
                status(value.status()),
                value.customerActorId(),
                value.agentActorId(),
                value.lastMessageSeq(),
                value.lastMessageAtUnixMs(),
                value.idleDeadlineUnixMs());
    }

    public static Messages.ChatMessage message(Conversation.Message value) {
        return new Messages.ChatMessage(
                value.conversationId(),
                value.messageSeq(),
                value.senderActorId(),
                value.text(),
                value.sentAtUnixMs());
    }

    public static String role(Conversation.Role value) {
        return value == Conversation.Role.Customer
                ? SampleNames.Roles.Customer
                : SampleNames.Roles.Agent;
    }

    private static String status(Conversation.Status value) {
        return switch (value) {
            case WaitingForAgent -> SampleNames.Statuses.WaitingForAgent;
            case Active -> SampleNames.Statuses.Active;
            case WaitingForClose -> SampleNames.Statuses.WaitingForClose;
            case Closed -> SampleNames.Statuses.Closed;
        };
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/ConversationContracts.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure

import systems.zlink.samples.kotlin.supportchat.server.configuration.ConversationStatuses
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationMessage
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationSnapshot
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationStatus
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ParticipantRole
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ChatMessage
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationState

object ConversationContracts {
    fun toState(snapshot: ConversationSnapshot): ConversationState =
        ConversationState(
            conversationId = snapshot.conversationId,
            subject = snapshot.subject,
            status = toStatus(snapshot.status),
            customerActorId = snapshot.customerActorId,
            agentActorId = snapshot.agentActorId,
            lastMessageSeq = snapshot.lastMessageSeq,
            lastMessageAtUnixMs = snapshot.lastMessageAtUnixMs,
            idleDeadlineUnixMs = snapshot.idleDeadlineUnixMs,
        )

    fun toMessage(message: ConversationMessage): ChatMessage =
        ChatMessage(
            conversationId = message.conversationId,
            messageSeq = message.messageSeq,
            senderActorId = message.senderActorId,
            text = message.text,
            sentAtUnixMs = message.sentAtUnixMs,
        )

    fun toStatus(status: ConversationStatus): String =
        when (status) {
            ConversationStatus.WaitingForAgent -> ConversationStatuses.WaitingForAgent
            ConversationStatus.Active -> ConversationStatuses.Active
            ConversationStatus.WaitingForClose -> ConversationStatuses.WaitingForClose
            ConversationStatus.Closed -> ConversationStatuses.Closed
        }

    fun toRole(role: ParticipantRole): String =
        when (role) {
            ParticipantRole.Customer -> SupportChatRoles.Customer
            ParticipantRole.Agent -> SupportChatRoles.Agent
        }
}
